README
NAME
    Mojolicious::Plugin::Cache::Action - Mojolicious plugin for caching

VERSION
    version 0.001

SYNOPSIS
    #Mojolicious $self->plugin('cache-action');

    #Mojolicious::Lite plugin 'cache-action';

DESCRIPTION
    This plugin caches responses of mojolicious applications. It works by
    caching the entire output of controller action for every request. Action
    caching internally uses the *before_dispatch* hook to serve the response
    from cache by skipping the entire controller body. Uncached responses
    are cached in an *after_dispatch* hook.

    The cache is named according to the current host and path. So, the cache
    will differentiate between an identical page that is accessed from
    tucker.myplace.com/user/2 and from caboose.myplace.com/user/2

    Different representation of the same resource such as
    tucker.myplace.com/book/list and tucker.myplace.com/book/list.json are
    considered as separate requests and so are cached separately.

  Cache backends
    This plugin uses CHI <http://http://search.cpan.org/~jswartz/CHI> for
    caching responses. So, all the various cache backends and customization
    options of CHI are supported. By default, this plugin uses the Memory
    <http://search.cpan.org/~jswartz/CHI-0.36/lib/CHI/Driver/Memory.pm>
    cache backend. Various other backends

    *   <http://search.cpan.org/~jswartz/CHI-0.36/lib/CHI/Driver/File.pm|Fil
        e>

    *   <http://search.cpan.org/~jswartz/CHI-0.36/lib/CHI/Driver/FastMmap.pm
        |FastMmap>

    *   <http://search.cpan.org/~jswartz/CHI-Driver-Memcached-0.12/lib/CHI/D
        river/Memcached.pm|Memcached>

    *   <http://search.cpan.org/~jswartz/CHI-Driver-BerkeleyDB-0.03/lib/CHI/
        Driver/BerkeleyDB.pm|BerkeleyDB>

    are also available through CHI.

  Options
    actions
         actions => [qw/action1 action2 ....]

         #Mojolicious::Lite 
         plugin caching-actions => { cache_actions => [qw/user show/]}; 

         By default,  all actions with successful GET requests will be cached

    options
          options =>  \%options
          All CHI module options are recognized

          #Mojolicious lite using memcache 
          plugin caching-actions => {
             cache_options => {
               driver => 'Memcached',  
               servers => [ "10.0.0.15:11211",  "10.0.0.15:11212" ] 
             }
          } 

          #Mojolicious lite using file based storage
          plugin caching-actions => {
             cache_options => {
               driver => 'File',  
               root_dir => '/path/to/cache' 
             }
          }

AUTHOR
    Siddhartha Basu <biosidd@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2010 by Siddhartha Basu.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
{{$NEXT}}

0.0017 04-14-2012

0.0017 04-12-2012

- Works with latest mojolicious

0.0015 04-25-2011

- Bumped mojolicious version to include its latest security patches.

0.0014 04-08-2011

- Added File::Path as dependency
- Fixed typos in the pod of action-cache plugin.

0.0013 02-25-2011

- First release



t/lib/book/lib/Book.pm
package Book;

use strict;
use base 'Mojolicious';
use File::Spec::Functions;
use Module::Build;

# This method will run once at server start
sub startup {
    my $self = shift;

    # Routes
    my $r = $self->routes;

    my $build = Module::Build->current;
    my $cache_dir = catdir( $build->base_dir, 't', 'tmp', 'cache' );

    $self->plugin( 'caching-actions',
        { cache_options => { driver => 'File', root_dir => $cache_dir } } );

    my $books = $r->route('/books')->to('controller-cache#books');
}

1;


t/lib/user/lib/User.pm
package User;

use strict;
use base 'Mojolicious';
use File::Spec::Functions;
use Module::Build;

# This method will run once at server start
sub startup {
    my $self = shift;

    # Routes
    my $r = $self->routes;

    my $build = Module::Build->current;
    my $cache_dir = catdir( $build->base_dir, 't', 'tmp', 'cache' );

    $self->plugin(
        'caching-actions',
        {   cache_options => {
                root_dir  => $cache_dir,
                namespace => 'user',
                driver    => 'File'
            },
            cache_actions => [qw/users show/]
        }
    );

    $r->route('/user')->via('post')->to('controller-cache#add');
    $r->route('/user/:id')->via('delete')->to('controller-cache#remove_user');
    my $books
        = $r->waypoint('/user')->via('get')->to('controller-cache#users');
    my $more
        = $books->waypoint('/:id')->via('get')->to('controller-cache#show');
    $more->route('/email')->via('get')->to('controller-cache#email');
    $more->route('/name')->via('get')->to('controller-cache#name');
}

1;


lib/Mojolicious/Plugin/Cache.pm
package Mojolicious::Plugin::Cache;

BEGIN {
    $Mojolicious::Plugin::Cache::VERSION = '0.0014';
}

# Module implementation
#
1;

# ABSTRACT: Mojolicious plugin for caching

__END__

=pod

=head1 NAME

Mojolicious::Plugin::Cache - Mojolicious plugin for caching

=head1 VERSION

version 0.0014

=head1 SYNOPSIS

Action caching:

  plugin 'cache-action';
  $self->plugin('cache-action');

Page caching:

 plugin 'cache-page'
 $self->plugin('cache-page');

=head1 DESCRIPTION

This distribution provides two plugins to cache responses of mojolicious applications,  action-caching
and page-caching. 

=over

=item *

L<Action caching|Mojolicious::Plugin::Cache::Action>

B<Action caching> works by caching the entire response of controller action for every
requrest. For more read L<here|Mojolicious::Plugin::Cache::Action>

=item *

L<Page caching|Mojolicious::Plugin::Cache::Page>

B<Page caching> works by caching the controller response in a HTML file which can be
served directly by webserver bypassing the controller altogether. For more read 
L<here|Mojolicious::Plugin::Cache::Page>

=back

=head1 AUTHOR

Siddhartha Basu <biosidd@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2010 by Siddhartha Basu.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Mojolicious/Plugin/Cache/Page.pm
package Mojolicious::Plugin::Cache::Page;


use strict;
use Carp;
use File::Path qw/make_path/;
use File::Spec::Functions;
use base qw/Mojolicious::Plugin/;

# Module implementation
#

sub register {
    my ( $self, $app, $conf ) = @_;

    my $home = $app->home;
    my $cache_dir
        = -e $app->home->rel_dir('public')
        ? $app->home->rel_dir('public')
        : $home->to_string;

    #if given as an option
    if ( defined $conf->{cache_directory} ) {
        $cache_dir = $conf->{cache_directory};
    }

    my $actions;
    if ( defined $conf->{actions} ) {
        $actions = { map { $_ => 1 } @{ $conf->{actions} } };
    }

    $app->hook(
        'after_dispatch' => sub {
            my ( $c ) = @_;

            ## - has to be GET request
            return if $c->req->method ne 'GET';

            ## - only successful response
            return if $c->res->code != 200;

            ## - only html response
            return if $c->res->headers->content_type !~ /^text\/html/;

            ## - have to match the action
            my $name = $c->stash('action');
            return
                if defined $conf->{actions}
                    and not exists $actions->{$name};

            my $parts = $c->req->url->path->parts;
            my $file_name;
            if ( @$parts == 1 ) {
                $file_name = catfile( $cache_dir, $parts->[0] . '.html' );
            }
            else {
                my $end = pop @$parts;
                my $folder = $cache_dir;
                $folder = catdir($folder, $_) for @$parts;
                make_path( $folder );
                $file_name = catfile( $folder, $end . '.html' );
            }

            $app->log->debug(
                "storing in cache for action **$name** in file $file_name");

            my $handler = IO::File->new( $file_name, 'w' )
                or croak "cannot create file:$!";
            $handler->print( $c->res->body );
            $handler->close;
        }
    );
    return;
}

1;

# ABSTRACT: Page caching plugin

=head1 SYNOPSIS

Mojolicious:

 $self->plugin('cache-page');

Mojolicious::Lite:

 plugin 'cache-page';


=head1 DESCRIPTION

This plugin caches the entire output of controller action in a HTML file which can be
delivered directly by the webserver without even going through the controller at all. The
cache will generate files for example user/2/view.html,  user/new.html etc which matches
the request urls. The webserver picks up the existing file by matching the path otherwise
it gets passed to the mojolicious controller.

The cache is named according to path and so the cache will not differentiate between 
an identical page that is accessed from B<tucker.myplace.com/user/2/view.html>
and from B<caboose.myplace.com/user/2/view.html>. This caching system ignores the query
parameters. So,  for example /users?page=1 will be cached as users.html and so another
call to /users?page=2 will retrieve the users.html. The cache is expired by deleting the
html file that defered it creation until a new request comes in. 


=head2 Options

=over

=item actions

 actions => [qw/action1 action2 ....]

 #Mojolicious::Lite 
 plugin cache-page => { actions => [qw/user show/]}; 

 By default,  all actions with successful GET requests will be cached
 
=item cache_directory

  By default, for mojolicious lite application the current working directory is set for
  page caching. For mojolicious application it is set to the B<public> folder. 

  #Mojolicious lite using memcache 
  plugin cache-page => { cache_direcotry => '/home/page_cache' }; 


=back




lib/Mojolicious/Plugin/ActionCache.pm
package Mojolicious::Plugin::Actioncache;

use strict;
use warnings;
use CHI;
use Carp;
use Scalar::Util qw/blessed/;
use base qw/Mojolicious::Plugin/;

# Module implementation
#

my $cache;
my $actions;

__PACKAGE__->attr( 'driver' => 'File' );
__PACKAGE__->attr('root_dir');

sub register {
    my ( $self, $app, $conf ) = @_;

    $self->root_dir( $app->home->rel_dir('cache') );

    if ( defined $conf->{cache_actions} ) {
        $actions = { map { $_ => 1 } @{ $conf->{cache_actions} } };
    }

    #setup cache
    if ( !$cache ) {
        if ( defined $conf->{cache_object} ) {
            croak "not a CHI object\n" if !blessed $conf->{cache_object};
            $cache = $conf->{cache_object};
        }
        elsif ( defined $conf->{cache_options} ) {
            my $opt = $conf->{cache_options};
            $opt->{driver} = $self->driver if not defined $opt->{driver};
            $cache = CHI->new(%$opt);
        }
        else {
            $cache = CHI->new(
                driver   => $self->driver,
                root_dir => $self->root_dir
            );
        }
    }

    if ( $app->log->level eq 'debug' ) {
        $cache->on_set_error('log');
        $cache->on_get_error('log');
    }

    $app->plugins->add_hook(
        'before_dispatch' => sub {
            my ( $self, $c ) = @_;
            my $path = $c->tx->req->url->to_abs->to_string;
            $app->log->debug( ref $path );
            if ( $cache->is_valid($path) ) {
                $app->log->debug("serving from cache for $path");
                my $data = $cache->get($path);
                $c->res->code( $data->{code} );
                $c->res->headers( $data->{headers} );
                $c->res->body( $data->{body} );
                $c->stash( 'from_cache' => 1 );
            }
        }
    );

    $app->plugins->add_hook(
        'after_dispatch' => sub {
            my ( $self, $c ) = @_;

            #conditions at which no caching will be done
            ## - it is already a cached response
            return if $c->stash('from_cache');

            ## - has to be GET request
            return if $c->req->method ne 'GET';

            ## - only successful response
            return if $c->res->code != 200;

            my $path = $c->req->url->to_abs->to_string;
            my $name = $c->stash('action');

            ## - have to match the action
            return
                if defined $conf->{cache_actions}
                    and not exists $actions->{$name};

            $app->log->debug("storing in cache for $path and action $name");
            $cache->set(
                $path,
                {   body    => $c->res->body,
                    headers => $c->res->headers,
                    code    => $c->res->code
                }
            );
        }
    );

    return;
}

1;

__END__

# ABSTRACT: Mojolicious plugin for caching response using CHI module


=head1 SYNOPSIS

#Mojolicious
 $self->plugin('actioncache');

#Mojolicious::Lite
  plugin 'actioncache';


lib/Mojolicious/Plugin/Caching/Actions.pm
package Mojolicious::Plugin::Caching::Actions;

use strict;
use warnings;
use CHI;
use Carp;
use Scalar::Util qw/blessed/;
use base qw/Mojolicious::Plugin/;

# Module implementation
#

my $cache;
my $actions;

__PACKAGE__->attr( 'driver' => 'Memory' );

sub register {
    my ( $self, $app, $conf ) = @_;

    if ( defined $conf->{cache_actions} ) {
        $actions = { map { $_ => 1 } @{ $conf->{cache_actions} } };
    }

    #setup cache
    if ( !$cache ) {
        if ( defined $conf->{cache_options} ) {
            my $opt = $conf->{cache_options};
            $opt->{driver} = $self->driver if not defined $opt->{driver};
            $cache = CHI->new(%$opt);
        }
        else {
            $cache = CHI->new(
                driver   => $self->driver
            );
        }
    }

    if ( $app->log->level eq 'debug' ) {
        $cache->on_set_error('log');
        $cache->on_get_error('log');
    }

    $app->plugins->add_hook(
        'before_dispatch' => sub {
            my ( $self, $c ) = @_;
            my $path = $c->tx->req->url->to_abs->to_string;
            $app->log->debug( ref $path );
            if ( $cache->is_valid($path) ) {
                $app->log->debug("serving from cache for $path");
                my $data = $cache->get($path);
                $c->res->code( $data->{code} );
                $c->res->headers( $data->{headers} );
                $c->res->body( $data->{body} );
                $c->stash( 'from_cache' => 1 );
            }
        }
    );

    $app->plugins->add_hook(
        'after_dispatch' => sub {
            my ( $self, $c ) = @_;

            #conditions at which no caching will be done
            ## - it is already a cached response
            return if $c->stash('from_cache');

            ## - has to be GET request
            return if $c->req->method ne 'GET';

            ## - only successful response
            return if $c->res->code != 200;

            my $path = $c->req->url->to_abs->to_string;
            my $name = $c->stash('action');

            ## - have to match the action
            return
                if defined $conf->{cache_actions}
                    and not exists $actions->{$name};

            $app->log->debug("storing in cache for $path and action $name");
            $cache->set(
                $path,
                {   body    => $c->res->body,
                    headers => $c->res->headers,
                    code    => $c->res->code
                }
            );
        }
    );

    return;
}

1;

__END__

# ABSTRACT: Mojolicious plugin for caching


=head1 SYNOPSIS

#Mojolicious
 $self->plugin('caching-actions');

#Mojolicious::Lite
  plugin 'caching-actions';


=head1 DESCRIPTION

This plugin caches responses of mojolicious applications. It works by caching the entire
output of controller action for every request that goes through it. Action caching internally uses
the I<before_dispatch> hook to serve the response if cache is present and the entire
controller body is skipped. If absent the output is cached in an I<after_dispatch> hook. 

The cache is named according to the current host and path. So,  the cache will
differentiate between an identical page that is accessed from B<tucker.myplace.com/user/2>
and from B<caboose.myplace.com/user/2> 

Different representation of the same resource such as B<tucker.myplace.com/book/list> and
B<tucker.myplace.com/book/list.json> are considered as separate requests and so are
cached separately.


=head2 Backend 

This plugin uses B<CHI> L<http://http://search.cpan.org/~jswartz/CHI> for caching responses.
So,  all the various cache backends and customization options are supported here. By
default,  this plugin uses the B<Memory> cache backend.

=head2 Options

=over

=item cache_actions

 cache_actions => [qw/action1 action2 ....]

 #Mojolicious::Lite 
 plugin caching-actions => { cache_actions => [qw/user show/]}; 

 By default,  all actions with successful GET requests will be cached
 
=item cache_options

  cache_options =>  \%options
  All CHI module options are recognized

  #Mojolicious lite using memcache 
  plugin caching-actions => {
     cache_options => {
       driver => 'Memcached',  
       servers => [ "10.0.0.15:11211",  "10.0.0.15:11212" ] 
     }
  } 

  #Mojolicious lite using file based storage
  plugin caching-actions => {
     cache_options => {
       driver => 'File',  
       root_dir => '/path/to/cache' 
     }
  } 


  By default,  the B<Memory> driver is used.

=back




lib/Mojolicious/Plugin/Cache/Action.pm
package Mojolicious::Plugin::Cache::Action;

use strict;
use warnings;
use CHI;
use Carp;
use Mojo::Base -base;
use base qw/Mojolicious::Plugin/;

# Module implementation
#

my ($actions, $cache);
has 'driver' => 'Memory';

sub register {
    my ( $self, $app, $conf ) = @_;

    if ( defined $conf->{actions} ) {
        $actions = { map { $_ => 1 } @{ $conf->{actions} } };
    }

    #setup cache
    if ( !$cache ) {
        if ( defined $conf->{options} ) {
            my $opt = $conf->{options};
            $opt->{driver} = $self->driver if not defined $opt->{driver};
            $cache = CHI->new(%$opt) ;
        }
        else {
            $cache = CHI->new( driver => $self->driver );
        }
    }

    if ( $app->log->level eq 'debug' ) {
        $cache->on_set_error('log');
        $cache->on_get_error('log');
    }

    $app->hook(
        'before_dispatch' => sub {
            my ($c)   = @_;
            my $path  = $c->req->url->to_abs->to_string;
            $app->log->debug( ref $path );
            if ( $cache->is_valid($path) ) {
                $app->log->debug("serving from cache for $path");
                my $data = $cache->get($path);
                $c->res->code( $data->{code} );
                $c->res->headers( $data->{headers} );
                $c->res->body( $data->{body} );
                $c->stash( 'from_cache' => 1 );
            }
        }
    );

    $app->hook(
        'after_dispatch' => sub {
            my ($c) = @_;

            #conditions at which no caching will be done
            ## - it is already a cached response
            return if $c->stash('from_cache');

            ## - has to be GET request
            return if $c->req->method ne 'GET';

            ## - only successful response
            return if $c->res->code != 200;

            my $path = $c->req->url->to_abs->to_string;
            my $name = $c->stash('action');

            ## - have to match the action
            return
                if defined $conf->{actions}
                    and not exists $actions->{$name};

            $app->log->debug("storing in cache for $path and action $name");
            $cache->set(
                $path,
                {   body    => $c->res->body,
                    headers => $c->res->headers,
                    code    => $c->res->code
                }
            );
        }
    );

    return;
}

1;

# ABSTRACT: Action caching plugin

=head1 SYNOPSIS

Mojolicious:

 $self->plugin('cache-action');

Mojolicious::Lite:

 plugin 'cache-action';


=head1 DESCRIPTION

This plugin caches responses of mojolicious applications. It works by caching the entire
output of controller action for every request. Action caching internally uses
the I<before_dispatch> hook to serve the response from cache by skipping the entire
controller body. Uncached responses are cached in an I<after_dispatch> hook. 

The cache is named according to the current host and path. So,  the cache will
differentiate between an identical page that is accessed from B<tucker.myplace.com/user/2>
and from B<caboose.myplace.com/user/2> 
Different representation of the same resource such as B<tucker.myplace.com/book/list> and
B<tucker.myplace.com/book/list.json> are considered as separate requests and so are
cached separately.


=head2 Cache backends 

This plugin uses L<CHI|http://search.cpan.org/~jswartz/CHI> for caching responses.
So,  all the various cache backends and customization options of B<CHI> are supported. By
default,  this plugin uses the 
L<Memory|http://search.cpan.org/~jswartz/CHI-0.36/lib/CHI/Driver/Memory.pm> cache backend.
Various other backends 

=over

=item *  

L<File|http://search.cpan.org/~jswartz/CHI-0.36/lib/CHI/Driver/File.pm> 

=item *

L<FastMmap|http://search.cpan.org/~jswartz/CHI-0.36/lib/CHI/Driver/FastMmap.pm>

=item *

L<Memcached|http://search.cpan.org/~jswartz/CHI-Driver-Memcached-0.12/lib/CHI/Driver/Memcached.pm>

=item *

L<BerkleyDB|http://search.cpan.org/~jswartz/CHI-Driver-BerkeleyDB-0.03/lib/CHI/Driver/BerkeleyDB.pm>

=back

are also available through CHI.

=head2 Options

=over

=item actions

 actions => [qw/action1 action2 ....]

 #Mojolicious::Lite 
 plugin cache-action => { actions => [qw/user show/]}; 


 #Mojolicious
 $self->plugin('cache-action' => { actions => [qw/user show/]});

 By default,  all actions with successful GET requests will be cached
 
=item options

  options =>  \%options
  All CHI module options are recognized

  #Mojolicious lite using memcache 
  plugin cache-action => {
       options => {
       driver => 'Memcached',  
       servers => [ "10.0.0.15:11211",  "10.0.0.15:11212" ] 
     }
  } 

  #Mojolicious lite using file based storage
  plugin cache-action => {
       options => {
       driver => 'File',  
       root_dir => '/path/to/cache' 
     }
  } 

  #Mojolicious using file based storage
  $self->plugin('cache-action' => { options => 
        { 
          driver => 'File',  
          root_dir => '/path/to/cache' 
        }
  });


=back


